resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,265 @@
+# Dart 2.0 Upper and Lower bounds
+
+leafp@google.com
+
+This documents the currently implemented upper and lower bound computation.
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+## Helper predicates
+
+The **TOP** predicate is true for any type which is in the equivalence class of
+top types.
+
+- **TOP**(`Object`) is true
+- **TOP**(`dynamic`) is true
+- **TOP**(`void`) is true
+- **TOP**(`FutureOr<T>`) is **TOP**(T)
+- **TOP**(T) is false otherwise
+
+The **BOTTOM** predicate is true for either of the two bottom types.
+
+- **BOTTOM**(`Null`) is true
+- **BOTTOM**(`bottom`) is true
+- **BOTTOM**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on the equivalence class of top
+types.
+
+- **MORETOP**(`void`, `T`) = true
+- **MORETOP**(`T`, `void`) = false
+- **MORETOP**(`dynamic`, `T`) = true
+- **MORETOP**(`T`, `dynamic`) = false
+- **MORETOP**(`Object`, `T`) = true
+- **MORETOP**(`T`, `Object`) = false
+- **MORETOP**(`FutureOr<T>`, `FutureOr<S>`) = **MORETOP**(T, S)
+
+## Upper bounds
+
+We define the upper bound of two types T1 and T2 to be **UP**(T1,T2) as follows.
+
+
+- **UP**(`T`, `T`) = `T`
+- **UP**(`T1`, `T2`) = `T1` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  **MORETOP**(`T1`, `T2`)
+- **UP**(`T1`, `T2`) = `T2` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  not **MORETOP**(`T1`, `T2`)
```

Comment:
I don't think this is true, because MORETOP(void, void) is true in either order.

Reply:
Ah, of course. Since it's a total order, that property only holds for all _distinct_ `T1`, `T2`. I believe the current wording is the clearest one we can have.